lib/Parser/Parser.cpp

```diff
@@ -203,6 +205,30 @@ Token Lexer::getNextToken(std::string &ErrStr) {
                         StringRef(NumBegin, NumEnd - NumBegin), 10)};
   }
 
+  if (*Begin == '(') {
+    ++Begin;
+    const char *NumBegin = Begin;
+    while (Begin != End && (*Begin == '0' || *Begin == '1' || *Begin == 'x')) {
+      ++Begin;
+    }
+    const char *NumEnd = Begin - 1;
+    if (Begin == NumBegin) {
+      ErrStr = "invalid knownbits string";
+      return Token{Token::Error, Begin, 0, APInt()};
+    }
+    Token T;
+    T.K = Token::KnownBits;
+    T.Pos = NumBegin;
+    T.Len = size_t(Begin - NumBegin);
+    T.PatternString = StringRef(NumBegin, NumEnd - NumBegin + 1); 
+    if (Begin != End && *Begin != ')') {
```

Comment:
Added a test case. Yes, actually you were right Begin != End is not
necessary since my validity checks keep track of each character one by one
and make sure that if something is wrong anywhere, it will give an error. I
have simplified this part now.

On Sat, Jul 18, 2015 at 10:27 PM, John Regehr <notifications@github.com>
wrote:

> In lib/Parser/Parser.cpp
> <https://github.com/google/souper/pull/195#discussion_r34954632>:
>
> > +    ++Begin;
> > +    const char *NumBegin = Begin;
> > +    while (Begin != End && (*Begin == '0' || *Begin == '1' || *Begin == 'x')) {
> > +      ++Begin;
> > +    }
> > +    const char *NumEnd = Begin - 1;
> > +    if (Begin == NumBegin) {
> > +      ErrStr = "invalid knownbits string";
> > +      return Token{Token::Error, Begin, 0, APInt()};
> > +    }
> > +    Token T;
> > +    T.K = Token::KnownBits;
> > +    T.Pos = NumBegin;
> > +    T.Len = size_t(Begin - NumBegin);
> > +    T.PatternString = StringRef(NumBegin, NumEnd - NumBegin + 1);
> > +    if (Begin != End && *Begin != ')') {
>
> please add a test case showing this
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/souper/pull/195/files#r34954632>.
>



-- 
Thanks and Regards,
Jubi Taneja